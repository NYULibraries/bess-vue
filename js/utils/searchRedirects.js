export const getitSearch = ({ institution, issn, title, searchType }) => {
  if (issn) {
    return `https://getit.library.nyu.edu/search/journal_search?&rfr_id=info%3Asid%2Fsfxit.com%3Acitation&rft.title=&rft.object_id=&umlaut.title_search_type=contains&rft.jtitle=${encodeURI(title)}&rft.issn=${encodeURI(issn)}&umlaut.institution=${institution}`;
  } else if (title) {
    return `https://getit.library.nyu.edu/search/journal_search?utf8=%E2%9C%93&umlaut.institution=${institution}&rfr_id=info%3Asid%2Fsfxit.com%3Acitation&rft.title=&rft.object_id=&rft.issn=&umlaut.title_search_type=${searchType}&rft.jtitle=${encodeURI(title)}&Generate_OpenURL2=Search`;
  }
    return `https://getit.library.nyu.edu/?umlaut.institution=${institution}`;
};

export const primoSearch = ({ tab, scope, bobcatUrl, search, institution, vid }) => {
  return `${bobcatUrl}/primo-explore/search?institution=${institution}&vid=${vid}&tab=${tab}&search_scope=${scope}&mode=basic&displayMode=full&bulkSize=10&highlight=true&dum=true&displayField=all&primoQueryTemp=${search}&query=any,contains,${search}&sortby=rank&lang=en_US`;
};
