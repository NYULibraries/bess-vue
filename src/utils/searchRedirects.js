const qsSortBy = ( orderArr ) => ( m, n ) => orderArr.indexOf( m ) - orderArr.indexOf( n );

const queryStringify = ( dict, { sort, encode = true } ) =>
    Object.keys( dict )
        .sort( sort )
        .reduce( ( res, k, idx, keys ) => {
            const v = dict[k];
            const noop = el => el
            const [ encodedKey, encodedValue ] = [ k, v ].map( encode ? encodeURIComponent : noop );
            const isNotLast = idx !== keys.length - 1;
            const queryString = v ? `${ encodedKey }=${ encodedValue }${ isNotLast ? '&' : '' }` : '';
            return `${ res }${ queryString }`;
        }, '' );

export const getitSearch = ( { institution, issn, title, type, getitUrl } ) => {
    const baseGetIt = `${ getitUrl }/search/journal_search?`;

    const staticParams = {
        rfr_id: 'info:sid/sfxit.com:citation',
    };
    const dynamicParams = {
        'umlaut.title_search_type': title ? type : undefined,
        'rft.jtitle'              : title,
        'rft.issn'                : issn,
        'umlaut.institution'      : institution,
    };
    const qsOrder = [
        'rfr_id',
        'umlaut.title_search_type',
        'rft.jtitle',
        'rft.issn',
        'umlaut.institution',
    ];
    const qsParams = queryStringify(
        { ...staticParams, ...dynamicParams },
        { sort: qsSortBy( qsOrder ) },
    );

    if ( issn || title ) {
        return `${ baseGetIt }${ qsParams }`;
    } else {
        return `${ getitUrl }/?umlaut.institution=${ institution }`;
    }
};

export const primoSearch = ( { tab, scope, bobcatUrl, search, institution, vid, searchMethod = 'search' } ) => {
    let qsParams;

    // Redirect to BobCat search if `search` is non-empty.
    // If `search` is empty of meaningful user input, redirect to the BobCat home
    // page instead of redirecting to a BobCat blank search, which returns error messages
    // that are potentially confusing.
    //
    // For details, see:
    //     https://nyu-lib.monday.com/boards/765008773/views/76580587/pulses/2183176320
    //     "Update bobcat embed to send users to bobcat homepage on blank search"
    if ( search.match( /\S+/ ) ) {
        const staticParams = {
            mode        : 'basic',
            displayMode : 'full',
            bulkSize    : '10',
            dum         : 'true',
            displayField: 'all',
            sortby      : 'rank',
            lang        : 'en_US',
        };

        const dynamicParams = {
            institution,
            vid,
            tab,
            search_scope  : scope,
            query         : `any,contains,${ encodeURIComponent( search ) }`,
            primoQueryTemp: encodeURIComponent( search ),
        };

        const qsOrder = [
            'institution',
            'vid',
            'tab',
            'search_scope',
            'mode',
            'displayMode',
            'bulkSize',
            'highlight',
            'dum',
            'displayField',
            'primoQueryTemp',
            'query',
            'sortby',
            'lang',
        ];
        qsParams = queryStringify(
            { ...staticParams, ...dynamicParams },
            { sort: qsSortBy( qsOrder ), encode: false },
        );
    } else {
        qsParams = `vid=${ vid }`;
    }

    return `${ bobcatUrl }/discovery/${ searchMethod }?${ qsParams }`;
};

export const guidesSearch = ( { search, guidesUrl } ) => {
    return `${ guidesUrl }/srch.php?&q=${ encodeURIComponent( search ) }`;
};

export default {
    guidesSearch,
    getitSearch,
    primoSearch,
};
