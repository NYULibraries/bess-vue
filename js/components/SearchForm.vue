<template>

  <primo-search-input
  v-if="engine==='primo'"
  :institution="institution"
  :searchKey="searchKey"
  ></primo-search-input>

  <form v-else v-on:submit.prevent="openSearch()">
    <div class="bobcat_embed_search_field">
      <getit-search-input
      v-if="engine==='getit'"
      :title="getitSearchValues.title"
      :issn="getitSearchValues.issn"
      :type="getitSearchValues.type"
      :typeOptions="typeOptions"
      @updateGetitForm="updateGetitForm"
      ></getit-search-input>

      <guides-search-input
      v-if="engine==='guides'"
      v-model="guideSearchValues.search"
      ></guides-search-input>

      <span class="bobat_embed_searchbox_submit_container">
        <input aria-label="Search" class="bobcat_embed_searchbox_submit" name="Submit" type="submit" value="GO">
      </span>
    </div>
  </form>
</template>

<script>
import { primoSearch, getitSearch, guidesSearch } from '../utils/searchRedirects';
import PrimoSearchInput from './PrimoSearchInput.vue';
import GetitSearchInput from './GetitSearchInput.vue';
import GuidesSearchInput from './GuidesSearchInput.vue';

export default {
  data() {
    return ({
      getitSearchValues: {
        title: '',
        issn: '',
        type: 'contains',
      },
      primoSearchValues: {
        search: '',
      },
      guideSearchValues: {
        search: '',
      }
    });
  },
  props: [
    'searchKey',
    'institution',
    'engine',
  ],
  components: {
    PrimoSearchInput,
    GetitSearchInput,
    GuidesSearchInput
  },
  computed: {
    typeOptions() {
      return CONFIG.institutions[this.institution].getitSearchValues[this.searchKey].searchTypes;
    },
    searchFunction() {
      return {
        primo: primoSearch,
        getit: getitSearch,
        guides: guidesSearch,
      }[this.engine];
    },
    searchValues() {
      return {
        primo: {
          ...this.primoSearchValues,
          // Bobcat search parameters
          ...CONFIG.institutions[this.institution].primoSearchValues[this.searchKey],
        },
        getit: {
          ...this.getitSearchValues
        },
        guides: {
          ...this.guideSearchValues
        },
      }[this.engine];
    },
  },
  methods: {
    openSearch() {
      const url = this.searchFunction({
        institution: this.institution,
        ...(this.searchValues),
      });

      window.open(url);
    },
    updateGetitForm(param, value) {
      if (this.getitSearchValues[param] !== undefined) {
        this.getitSearchValues[param] = value;
      }
    }
  },
};
</script>

<style lang="scss">
</style>
