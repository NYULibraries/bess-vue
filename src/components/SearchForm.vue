<template>
  <search-redirect-form
    :search-key="searchKey"
    :search-function="searchFunction"
    :search-engine-props="engine"
    :input-aria-label="inputAriaLabel"
    :ui="ui"
  />
</template>

<script>
import SearchRedirectForm from '@/components/searchForms/SearchRedirectForm.vue';
import { primoSearch, guidesSearch } from '@/utils/searchRedirects';

export default {
    components: {
        SearchRedirectForm,
    },
    props: {
        engine: {
            type   : String,
            default: undefined,
        },
        searchKey: {
            type    : String,
            required: true,
        },
        ui: {
            type   : Object,
            default: {},
        },
    },
    computed: {
        engineType() {
            return this.engine?.type;
        },
        searchFunction() {
            const fxns = {
                primo : primoSearch,
                guides: guidesSearch,
            };

            return fxns[ this.engineType ];
        },
        inputAriaLabel() {
            const labels = {
                primo : 'Search Bobcat',
                guides: 'Search for research guides',
            }

            return labels[ this.engineType ];
        },
    },
};
</script>
