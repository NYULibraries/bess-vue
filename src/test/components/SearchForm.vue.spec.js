// Note on disabled tests:
//
// `shallowMount()` is supposed to stub out child components, but there appears
// to be a bug which causes the read child components to be used when
// `NODE_ENV=production` is set.  We are now running tests in production mode
// by default, both because it makes sense for this project and  Node team
// strongly recommends it:
//    https://nodejs.org/en/learn/getting-started/nodejs-the-difference-between-development-and-production
// TODO: Determine if this is a bug or not, and if so, report it or submit a PR,
//       and/or work around this issue.

import { beforeEach, describe, expect, test } from 'vitest';
import { guidesSearch, primoSearch } from '../../utils/searchRedirects';

import SearchForm from '../../components/SearchForm.vue';
import { shallowMount } from '@vue/test-utils';

import appConfig from '../../../config';

describe( 'SearchForm', () => {
    let currentFormConfig;
    let currentProps;
    let wrapper;

    describe.each(
        Object.keys( appConfig.institutions ).map(
            function ( institution ) {
                return { institution };
            },
        ) )( '$institution', ( { institution } ) => {
        beforeEach( () => {
            currentFormConfig = appConfig.institutions[ institution ][ 0 ];

            currentProps = {
                searchKey: 1,
                engine   : currentFormConfig.engine,
                ui       : currentFormConfig.ui || {},
            };

            wrapper = shallowMount( SearchForm, {
                props: currentProps,
            } );
        } );

        test( 'is a Vue instance', () => {
            expect( wrapper ).toBeTruthy();
        } );

        // See "Note on disabled tests" at the top of this file.
        // The snapshots show the `SearchRedirectForm` child component being
        // rendered instead of stubbed.  The snapshot presumably has the correct
        // HTML, but once the presumed bug is fixed, this test will fail since
        // the `SearchRedirectForm` child component will once again be stubbed.
        // test( 'has the correct HTML', () => {
        //     expect( wrapper.html() ).toMatchSnapshot();
        // } );

        describe( 'props', () => {
            test( 'includes searchKey and engine', () => {
                expect( Object.keys( wrapper.props() ).length ).toEqual( 3 );
                expect( wrapper.props().searchKey ).toEqual( currentProps.searchKey );
                expect( wrapper.props().engine ).toEqual( currentProps.engine );
            } );
        } );

        describe( 'computed', () => {
            describe( 'engineType', () => {
                test( 'evaluates engine type', () => {
                    expect( wrapper.vm.engineType ).toBe( currentProps.engine.type );
                } );

                test( 'is undefined if engine is undefined', async () => {
                    await wrapper.setProps( { engine: undefined } );
                    expect( wrapper.vm.engineType ).toBe( undefined );
                } );
            } );

            describe( 'searchFunction', () => {
                test( 'is properly mapped to engines', async () => {
                    expect( wrapper.vm.searchFunction ).toBe( primoSearch );
                    await wrapper.setProps( { engine: { type: 'guides', scopesMap: {}, scope: '' } } );
                    expect( wrapper.vm.searchFunction ).toBe( guidesSearch );
                    await wrapper.setProps( { engine: { type: 'unknown' } } );
                    expect( wrapper.vm.searchFunction ).toBeUndefined();
                } );
            } );

            describe( 'inputAriaLabel', () => {
                test( 'has appropriate label for \'primo\' engine', () => {
                    expect( wrapper.vm.inputAriaLabel ).toBe( 'Search Bobcat' );
                } );
                test( 'has appropriate label for \'guides\' engine', async () => {
                    await wrapper.setProps( { engine: { type: 'guides', scopesMap: {}, scope: '' } } );
                    expect( wrapper.vm.inputAriaLabel ).toBe( 'Search for research guides' );
                } );
                test( 'is empty for an unknown engine', async () => {
                    await wrapper.setProps( { engine: { type: 'unknown' } } );
                    expect( wrapper.vm.inputAriaLabel ).toBeUndefined();
                } );
            } );

            describe( 'ui.searchScopeDropdown', () => {
                test( 'returns the `ui.searchScopeDropdown` from the config', () => {
                    expect( wrapper.vm.ui.searchScopeDropdown ).toEqual( currentProps.ui.searchScopeDropdown );
                } );

                test( 'is `undefined` if `ui.searchScopeDropdown` is not defined', async () => {
                    await wrapper.setProps( { ui: {} } );
                    expect( wrapper.vm.ui.searchScopeDropdown ).toBeUndefined();
                } );
            } );

            // See "Note on disabled tests" at the top of this file.
            // describe( 'shallow render', () => {
            //     test( 'renders <search-redirect-form> if primo or guides engine', async () => {
            //         expect( wrapper.find( 'search-redirect-form-stub' ).exists() ).toBeTruthy();
            //         await wrapper.setProps( { engine: { type: 'guides' } } );
            //         expect( wrapper.find( 'search-redirect-form-stub' ).exists() ).toBeTruthy();
            //         expect( wrapper.find( 'getit-search-form-stub' ).exists() ).toBeFalsy();
            //     } );
            // } );
        } );
    } );
} );
