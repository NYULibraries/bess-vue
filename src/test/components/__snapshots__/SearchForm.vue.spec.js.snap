// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > has the correct HTML 1`] = `"<search-redirect-form-stub searchkey="test" searchfunction="[Function]" searchengineprops="[object Object]" inputarialabel="Search Bobcat" scopesconfig="[object Object]" enginescope="scope1"></search-redirect-form-stub>"`;
